lib/source-map/mapping-list.js

```diff
@@ -0,0 +1,88 @@
+/* -*- Mode: js; js-indent-level: 2; -*- */
+/*
+ * Copyright 2014 Mozilla Foundation and contributors
+ * Licensed under the New BSD license. See LICENSE or:
+ * http://opensource.org/licenses/BSD-3-Clause
+ */
+if (typeof define !== 'function') {
+    var define = require('amdefine')(module, require);
+}
+define(function (require, exports, module) {
+
+  var util = require('./util');
+
+  /**
+   * Determine whether mappingB is after mappingA with respect to generated position.
+   */
+  function generatedPositionAfter(mappingA, mappingB) {
+    // Optimized for most common case
+    var lineA = mappingA.generatedLine;
+    var lineB = mappingB.generatedLine;
+    var columnA = mappingA.generatedColumn;
+    var columnB = mappingB.generatedColumn;
+    return lineB > lineA || lineB == lineA && columnB >= columnA ||
+           util.compareByGeneratedPositions(mappingA, mappingB) <= 0;
+  }
+
+  /**
+   * A data structure to provide a sorted view of accumulated mappings in a
+   * performance conscious manner. It trades a neglibable overhead in general
+   * case for a large speedup in case of mappings being added in order.
+   */
+  function MappingList() {
+    this._array = [];
+    this._unordered = false;
+    // Serves as infimum
+    this._last = {generatedLine: -1, generatedColumn: 0};
+  }
+
+  /**
+   * Static method for creating MappingList instances from an existing array.
+   */
+  MappingList.fromArray = function MappingList_fromArray(aArray) {
+    var list = new MappingList();
+    list._array = aArray.slice().sort(util.compareByGeneratedPositions);
+    return list;
+  };
+
+
+  /**
+   * Iterate through internal items.
+   */
+  MappingList.prototype.forEach = function MappingList_forEach() {
+    return Array.prototype.forEach.apply(this._array, arguments);
+  };
```

Comment:
So, it does get used here https://github.com/crisptrutski/source-map/blob/optimistic-add-mapping/lib/source-map/source-map-generator.js#L190. 

Used to make local mutations to the elements, so no cares about order. Could skip return value in wrapper to obliviate one self-harm vector, but yeah it'd be confusing with functions with non-scoped side effects.

To be honest I don't see much scope for reuse of this class though - passing in comparator as function was too much overhead so now it's really tightly coupled and just an implementation detail of the generator.

How about an ignoble `this._mappings._array.forEach` there and killing the wrapper method? You could consider `_array` to be in "friend" scope :goat: